openedx/core/lib/gating/api.py

```diff
@@ -299,3 +304,105 @@ def get_gated_content(course, user):
                 {'id': user.id}
             )
         ]
+
+
+def is_prereq_met(content_id, user_id, recalc_on_unmet=False):
+    """
+    Returns true if the prequiste has been met for a given milestone
+
+    Arguments:
+        content_id (BlockUsageLocator): BlockUsageLocator for the content
+        user_id: The id of the user
+        recalc_on_unmet: Recalculate the grade if prereq has not yet been met
+
+    Returns:
+        tuple: True|False,
+        prereq_meta_info = { 'url': prereq_url, 'display_name': prereq_name}
+    """
+    course_id = content_id.course_key
+
+    # if unfullfilled milestones exist it means prereq has not been met
+    unfulfilled_milestones = milestones_helpers.get_course_content_milestones(
+        course_id,
+        content_id,
+        'requires',
+        user_id
+    )
+    prereq_met = not unfulfilled_milestones
+
+    if prereq_met or not recalc_on_unmet:
+        return prereq_met, {}
+
+    milestone = unfulfilled_milestones[0]
+    student = User.objects.get(id=user_id)
+    store = modulestore()
+
+    with store.bulk_operations(course_id):
+        course_structure = get_course_blocks(student, store.make_course_usage_key(course_id))
+        course = store.get_course(course_id, depth=0)
+        subsection_grade_factory = SubsectionGradeFactory(student, course, course_structure)
+        subsection_usage_key = UsageKey.from_string(milestone['namespace'].replace(GATING_NAMESPACE_QUALIFIER, ''))
+
+        if subsection_usage_key in course_structure:
+            # this will force a recalcuation of the subsection grade
+            subsection_grade = subsection_grade_factory.update(course_structure[subsection_usage_key])
+            prereq_met = update_milestone(milestone, subsection_grade, milestone, user_id)
+            prereq_meta_info = {
+                'url': reverse('jump_to', kwargs={'course_id': course_id, 'location': subsection_usage_key}),
+                'display_name': store.get_item(subsection_usage_key).display_name
+            }
+
```

Comment:
@nasthagiri @ormsbee 
Please take a look at this function as it is where I force the subsection grade recalculation. Want to make sure I'm doing it in an optimal way. Didn't know how expensive the call to get_course_blocks() is, but seems it's needed in order to instantiate a SubsectionGradeFactory.
Also please note I have written tests yet, wanted to first get general approval for approach as Dave and I discussed.

Reply:
It seems reasonable to me, but I definitely defer to @nasthagiri on matters related to grade calculation.

Reply:
@bill-filler Thanks for flagging this.  Yes, there is a cost in calling `get_course_blocks` and you can expect it to take 500ms for large courses with many blocks.  For your case, since you don't need all the blocks in the entire course, it would be great if you could try getting the blocks for only the subsection in question.  The parameter names in the `SubsectionGrade` APIs assume it's the entire course_structure.  However, a quick read of the code indicates that it should still work if only the subsection's structure is passed in.
I'll add specific comments inline with the code here.

Reply:
updated per suggestions

Reply:
thanks.  Were you able to manually test it to confirm it works with the `subsection_structure`?

Reply:
Yes I manually confirmed it works as expected with subsection_structure